more python.d.plugin debugging

ktsaou opened the pull request:
@paulfantom I have added the timing info I was looking for.

I have also added 3 FIXME comments I would like you to check.

paulfantom commented on plugins.d/python.d.plugin:
```diff
@@ -255,6 +255,7 @@ class PythonCharts(object):
         defaults = {}
         for key in BASE_CONFIG:
+            # FIXME for some reason this is called 3 times per module
             debug(module.__name__ + ": reading configuration")
```
`debug()` should be before `for`. This loop looks and sets up default module configuration (`update_every`, `retries`, `priority`)

paulfantom commented on plugins.d/python.d.plugin:
```diff
@@ -515,6 +519,9 @@ def parse_cmdline(directory, *commands):
             mods.append(cmd.replace(".chart.py", ""))
         else:
             try:
+                # FIXME for some reason this overwrites the module configuration
```
replace `else try` to:
```
if DEBUG_FLAG:
    for cmd in commands[1:]:
        try:   

```

ktsaou replied:
This might not be right. By default, python modules configuration should not have an `update_every`. So, the environment variable `NETDATA_UPDATE_EVERY` and/or the command line option should set the default.

paulfantom replied:
But this is only for debugging when invoking `python.d.plugin debug 1`. If you don't pass `debug some_number` everything works as expected.

paulfantom replied:
I have probably misunderstood you. And I have a better solution than previous one:
replace
```py
else:
    try:
        BASE_CONFIG['update_every'] = int(cmd)
        OVERRIDE_UPDATE_EVERY = True
        debug(PROGRAM, "overriding update interval to", str(int(cmd)))
    except ValueError:
        pass

```
```py
for cmd in commands[1:]:
    try:
        BASE_CONFIG['update_every'] = int(cmd)
        if DEBUG_FLAG:
            OVERRIDE_UPDATE_EVERY = True
            debug(PROGRAM, "overriding update interval to", str(int(cmd)))
    except ValueError:
        pass
```

paulfantom commented on plugins.d/python.d.plugin:
```diff
@@ -364,6 +365,8 @@ class PythonCharts(object):
                 else:
+                    # FIXME job.name is incomplete here
+                    # it shows None is example and only the job name without the module in mysql
                     debug(job.name, ": check succeeded")
```
change `job.name` to `job.chart_name`

paulfantom commented on python.d/pip_install.sh:
```diff
@@ -1,5 +1,9 @@
+
 PIP=`which pip`
 
 ${PIP} install --target="python_modules" yaml
```
@1 It just needs to check if there is `python_modules` directory and even if there isn't, it won't work. But you can add something like this after shebang:
```sh
[ -d 'python_modules' ] || ( echo "Cannot find 'python_modules' directory, exiting"; exit 1)
```

@2 In Python 3 packages are named differently, and there is no `mysqldb` in pypi repository.

```sh
${PIP} install --target="python_modules" yaml || ${PIP} install --target="python_modules" PyYAML
${PIP} install --target="python_modules" mysqldb || ${PIP} install --target="python_modules" pymysql
```

paulfantom commented:
Added comments to every FIXME. If you apply them, those issues should be resolved.

ktsaou commented:
ok. I did all the changes and also updated the python installer.

pip has a bug in certain systems (i.e. fedora24) and cannot install anything when `--target` is given. So, I have added an option to the installer to install the modules system-wide.

Take a look. If you also believe it is good let me know to merge this.

ktsaou commented:
Regarding the pip bug: https://github.com/pypa/pip/issues/3056

paulfantom commented on plugins.d/python.d.plugin:
```diff
@@ -516,12 +516,10 @@ def parse_cmdline(directory, *commands):
-                debug(PROGRAM, "overriding update interval to", str(int(cmd)))
+                if DEBUG_FLAG:
+                    OVERRIDE_UPDATE_EVERY = True
+                    debug(PROGRAM, "overriding update interval to", str(int(cmd)))
```
You cannot do it like this. What about:
`python.d.plugin 1 debug`?

ktsaou replied:
You are right.
Would you solve this after merge?

paulfantom replied:
I solved it in previous note:)
But merge it and I will fix this.

paulfantom commented on plugins.d/python.d.plugin:
```diff
@@ -418,13 +418,14 @@ class PythonCharts(object):
         # check if it is time to execute job update() function
         if job.timetable['next'] > t_start:
-            debug("it is not a time to invoke update on", job.chart_name)
+            debug(job.chart_name + " will be run in " + str(int((job.timetable['next'] - t_start) * 1000)) + " ms")
```
Just a note: this `if` is just a precaution, it should never get in here.

ktsaou replied:
```
python.d DEBUG:  created charts for example
python.d DEBUG:  created charts for mysql_ptcp
python.d DEBUG:  example will be run in 1593 ms
python.d DEBUG:  mysql_ptcp will be run in 1593 ms
python.d DEBUG:  example updated in 0 ms
python.d DEBUG:  mysql_ptcp ready to run, after 1626 ms (update_every: 5000 ms, latency: 0 ms)
python.d DEBUG:  mysql_ptcp updated in 2 ms
```

well, it is used

paulfantom replied:
Oh, so I need to be optimize it.

paulfantom commented:
Wow. I see you changed my simple helper script into fully fledged installer script. Nice.

paulfantom commented:
Just one note on the installer. 
We don't need packages in `python_modules` when they are installed system-wide. So if someone has installed needed packages this script will install them for nothing.

ktsaou commented:
> We don't need packages in python_modules when they are installed system-wide. So if someone has installed needed packages this script will install them for nothing.

Yes, I know. But the same script can install them system-wide with `-s` for those that need help.

paulfantom commented:
OK
I think it is ready to be merged.

ktsaou commented:
merged. Thanks!
